pycrires/pipeline.py

```diff
@@ -3726,108 +3726,90 @@ def correct_wavelengths(self, nod_ab: str = "A") -> None:
-                )
+                print(f"\nCorrecting wavelength solution of detector {i_det+1}:")
 
-                print(
-                    f"   - {spec_name} -> lambda = {opt_result.x[0]:.4f} "
-                    f"+ {opt_result.x[1]:.1f} * lambda'"
-                )
-
-                if not opt_result.success:
-                    warnings.warn(
-                        f"Could not converge with correcting the "
-                        f"wavelength solution of {spec_name[:-3]}."
+                # Loop over spectral orders
+                for spec_name in spec_orders:
+                    # Extract WL, SPEC, and ERR for given order/detector
+                    wavel = hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_WL"]
+                    spec = hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_SPEC"]
+                    err = hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_ERR"]
+
+                    #Remove continuum and nans of spectra
+                    nans = np.isnan(spec)
+                    continuum = ndimage.gaussian_filter1d(spec[~nans], 100, mode='reflect')
+                    spec_flat = spec[~nans]/continuum - 1
+
+                    #Don't use the edges as that sometimes gives problems
+                    spec_flat = spec_flat[20:-20]
+                    used_wavel = wavel[~nans][20:-20]
+                    
+                    #Prepare cross-correlation grid
+                    a_grid = np.linspace(0.9, 1.1, 100)[:,np.newaxis, np.newaxis]
+                    b_grid = np.linspace(-0.5,0.5, 100)[np.newaxis,:, np.newaxis]
+                    mean_wavel = np.mean(wavel)
+                    wl_matrix = a_grid * (used_wavel[np.newaxis,np.newaxis,:] - mean_wavel) + mean_wavel + b_grid
+                    template = transm_interp(wl_matrix) - 1
+                    template_std = np.mean(np.std(template, axis=-1))
+
+                    #Check if there are enough telluric features in this wavelength range
+                    if template_std>0.01:
+                        #cross-correlation
+                        cross_corr = template.dot(spec_flat)
+
+                        #Find optimal wavelength solution
+                        opt_idx = np.unravel_index(np.argmax(cross_corr), cross_corr.shape)
+                        opt_a = a_grid[opt_idx[0],0,0]
+                        opt_b = b_grid[0,opt_idx[1],0]
+                    else:
+                        opt_a = 1.
+                        opt_b = 0.
+                    
+                    print(
+                        f"   - {spec_name} -> lambda = {opt_b:.4f} "
+                        f"+ {opt_a:.4f} * lambda'"
                     )
 
-                hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_WL"] = (
-                    opt_result.x[0] + opt_result.x[1] * spec_data[:, 0]
-                )
+                    hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_WL"] = (
```

Comment:
Should the last part be `+ mean_wavel + opt_b`?

Reply:
Yes indeed, no clue why I had a  minus sign there.